src/librustc/ty/structural_impls.rs

```diff
@@ -1073,3 +1060,29 @@ impl<'tcx> TypeFoldable<'tcx> for InferConst<'tcx> {
         false
     }
 }
+
+fn fold_list<'tcx, F, T>(
```

Comment:
Might be worth adding a comment here that explain that this is an optimization over `.iter().map(...).collect()`

Reply:
Done in https://github.com/rust-lang/rust/pull/68031/commits/a1586f1d2b17d687444d3b94aedd7ce24ae074ed